planner/core/common_plans.go

```diff
@@ -727,7 +727,12 @@ func (e *Execute) buildRangeForTableScan(sctx sessionctx.Context, ts *PhysicalTa
 		for _, colInfo := range pk.Columns {
 			if pkCol := expression.ColInfo2Col(ts.schema.Columns, ts.Table.Columns[colInfo.Offset]); pkCol != nil {
 				pkCols = append(pkCols, pkCol)
-				pkColsLen = append(pkColsLen, colInfo.Length)
+				// We need to consider the prefix index.
+				if colInfo.Length != types.UnspecifiedLength && colInfo.Length == pkCol.RetType.Flen {
+					pkColsLen = append(pkColsLen, types.UnspecifiedLength)
+				} else {
+					pkColsLen = append(pkColsLen, colInfo.Length)
+				}
```

Comment:
I'm not sure this can cover all the cases. This processing borrows the ideas from the three functions of `DataSource.deriveCommonHandleTablePathStats`, `expression.IndexInfo2PrefixCols` and `expression.IndexInfo2Cols`. Need more investigation to confirm.

Reply:
How about adding this comment to the file?